memcached.c

```diff
@@ -4181,6 +4263,740 @@ static inline void process_get_command(conn *c, token_t *tokens, size_t ntokens,
+    olen = tokens[KEY_TOKEN + 1].length;
+    rtokens--;
+
+    if (olen > MFLAG_MAX_OPT_LENGTH) {
+        out_string(c, "CLIENT_ERROR options flags are too long");
+        return;
+    }
+
+    // Copy opts into wbuf and advance pointer.
+    // We return the initial options + extra indicator flags.
+    // we reserve 4 bytes in front of the buffer, for up to three extra flags
+    // we can tag plus the initial space.
+    // This could be simpler by adding two iov's for the header line, but this
+    // is a hot path so trying to keep those to a minimum.
+    fp = c->wbuf + 4;
+    memcpy(fp, opts, olen);
+    p = fp + olen;
+    fp--; // next token, or final space, goes here.
+
+    // scrubs duplicated options and sets flags for how to load the item.
+    rtokens -= _meta_flag_preparse(opts, olen, &of);
+    c->noreply = of.no_reply;
+
+    // FIXME: make string more clear and add key to response
+    if (rtokens < 0) {
+        out_string(c, "CLIENT_ERROR not enough tokens supplied");
+        return;
+    }
+    rtokens = KEY_TOKEN + 2;
+
+    // TODO: need to indicate if the item was overflowed or not?
+    // I think we do, since an overflow shouldn't trigger an alloc/replace.
+    if (!of.locked) {
+        it = limited_get(key, nkey, c, 0, false, update);
+    } else {
+        it = limited_get_locked(key, nkey, c, update, &hv);
+    }
+
+    if (it == NULL && of.vivify) {
+        // Fill in the exptime during parsing later.
+        it = item_alloc(key, nkey, 0, realtime(0), 2);
+        // We don't actually need any of do_store_item's logic:
+        // - already fetched and missed an existing item.
+        // - lock is still held.
+        // - not append/prepend/replace
+        // - not testing CAS
+        if (it != NULL) {
+            // I look forward to the day I get rid of this :)
+            memcpy(ITEM_data(it), "\r\n", 2);
+            // NOTE: This initializes the CAS value.
+            do_item_link(it, hv);
+            item_created = true;
+        }
+    }
+
+    // don't have to check result of add_iov() since the iov size defaults are
+    // enough.
+    if (it) {
+        int32_t exptime_int = 0;
+
+        // Has this item already sent a token?
+        // Important to do this here so we don't send W with Z.
+        // Isn't critical, but easier for client authors to understand.
+        if (it->it_flags & ITEM_TOKEN_SENT) {
+            *fp = 'Z';
+            fp--;
+        }
+        if (it->it_flags & ITEM_STALE) {
+            *fp = 'X';
+            fp--;
+            // FIXME: think hard about this. is this a default, or a flag?
+            if ((it->it_flags & ITEM_TOKEN_SENT) == 0) {
+                // If we're stale but no token already sent, now send one.
+                won_token = true;
+            }
+        }
+
+        for (i = 0; i < olen; i++) {
+            switch (opts[i]) {
+                case 'T':
+                    // FIXME: I broke this.
```

Comment:
Looks like we need to fix this one to get the `mget foo T <ttl>` working.

Reply:
As the comment says, yes :)